diff --git a/clang/lib/Sema/SemaExpr.cpp b/clang/lib/Sema/SemaExpr.cpp
index 45616dadcbee4a86e5b0c60b4a756c8ba4dfd1c7..b7f1b3e875bf582c2f279cf4b0a8af219f1a1a9b 100644
--- a/clang/lib/Sema/SemaExpr.cpp
+++ b/clang/lib/Sema/SemaExpr.cpp
@@ -16520,10 +16520,10 @@ void Sema::PopExpressionEvaluationContext() {
 
   if (!Rec.Lambdas.empty()) {
     using ExpressionKind = ExpressionEvaluationContextRecord::ExpressionKind;
-    if (Rec.ExprContext == ExpressionKind::EK_TemplateArgument || Rec.isUnevaluated() ||
+    if (Rec.ExprContext == ExpressionKind::EK_TemplateArgument || (Rec.isUnevaluated() && !getLangOpts().CPlusPlus20) ||
         (Rec.isConstantEvaluated() && !getLangOpts().CPlusPlus17)) {
       unsigned D;
-      if (Rec.isUnevaluated()) {
+      if (Rec.isUnevaluated() && !getLangOpts().CPlusPlus20) {
         // C++11 [expr.prim.lambda]p2:
         //   A lambda-expression shall not appear in an unevaluated operand
         //   (Clause 5).
